File: services/ai/pipeline.ts
import { z } from 'zod';

// Define input/output schemas for each phase
const UnderstandInputSchema = z.object({
  userMessage: z.string(),
  context: z.object({
    annotations: z.array(z.any()),
    entryContent: z.string(),
    entryType: z.string(),
  }),
});

const UnderstandOutputSchema = z.object({
  intent: z.string(),
  entities: z.record(z.any()),
  complexity: z.number(),
  ambiguity: z.number(),
  requiredActions: z.array(z.string()),
});

const ReasonInputSchema = z.object({
  understandOutput: UnderstandOutputSchema,
  context: z.object({
    annotations: z.array(z.any()),
    entryContent: z.string(),
    entryType: z.string(),
  }),
});

const ReasonOutputSchema = z.object({
  plan: z.object({
    steps: z.array(z.string()),
    reasoning: z.string(),
  }),
  confidence: z.number(),
});

const ActInputSchema = z.object({
  reasonOutput: ReasonOutputSchema,
  understandOutput: UnderstandOutputSchema,
  context: z.object({
    annotations: z.array(z.any()),
    entryContent: z.string(),
    entryType: z.string(),
  }),
});

const ActOutputSchema = z.object({
  response: z.string(),
  actions: z.array(z.object({
    type: z.string(),
    payload: z.any(),
  })),
  version: z.literal('cognition.v1'),
  diagnostics: z.any().optional(),
});

// Complexity scoring function
export function complexity_score(input: { userMessage: string; context?: any }): number {
  const tokens = input.userMessage.split(/\s+/).length;
  const ambiguity = (input.userMessage.match(/\?/g) || []).length + (input.userMessage.match(/\b(or|maybe|perhaps|might|could|possibly)\b/gi) || []).length;
  const requiredActions = (input.userMessage.match(/\b(create|schedule|plan|analyze|reflect)\b/gi) || []).length;

  // Simple weighted score: tokens/100 + ambiguity*0.5 + requiredActions
  return (tokens / 100) + (ambiguity * 0.5) + requiredActions;
}

// Phase 1: Understand
export async function understand(input: z.infer<typeof UnderstandInputSchema>): Promise<z.infer<typeof UnderstandOutputSchema>> {
  // Placeholder implementation - would integrate with actual NLP/intent classification
  const complexity = complexity_score({ userMessage: input.userMessage });
  const ambiguity = Math.min(complexity * 0.2, 1);
  const intent = 'reflection'; // Placeholder
  const entities = {};
  const requiredActions = ['analyze', 'respond']; // Placeholder

  return UnderstandOutputSchema.parse({
    intent,
    entities,
    complexity,
    ambiguity,
    requiredActions,
  });
}

// Phase 2: Reason
export async function reason(input: z.infer<typeof ReasonInputSchema>): Promise<z.infer<typeof ReasonOutputSchema>> {
  // Placeholder implementation - would use reasoning engine
  const plan = {
    steps: ['Analyze user intent', 'Generate response', 'Check coherence'],
    reasoning: 'Based on understanding phase, create a structured plan',
  };
  const confidence = 0.8;

  return ReasonOutputSchema.parse({
    plan,
    confidence,
  });
}

// Phase 3: Act
export async function act(input: z.infer<typeof ActInputSchema>): Promise<z.infer<typeof ActOutputSchema>> {
  // Placeholder implementation - would generate final response
  const response = 'This is a placeholder response from the cognition engine.';
  const actions: Array<{ type: string; payload: unknown }> = [];

  return ActOutputSchema.parse({
    response,
    actions,
    version: 'cognition.v1',
  });
}

// Main pipeline orchestrator
export async function runPipeline(input: {
  userMessage: string;
  context: {
    annotations: any[];
    entryContent: string;
    entryType: string;
  };
}): Promise<z.infer<typeof ActOutputSchema>> {
  // Validate input
  const validatedInput = UnderstandInputSchema.parse(input);

  // Phase 1: Understand
  const understandResult = await understand(validatedInput);

  // Phase 2: Reason
  const reasonInput = ReasonInputSchema.parse({
    understandOutput: understandResult,
    context: input.context,
  });
  const reasonResult = await reason(reasonInput);

  // Phase 3: Act
  const actInput = ActInputSchema.parse({
    reasonOutput: reasonResult,
    understandOutput: understandResult,
    context: input.context,
  });
  const actResult = await act(actInput);

  return actResult;
}

File: services/python-ai/app/ai_legacy/pipeline.py
from typing import Any, Dict, List

from pydantic import BaseModel


class UnderstandInput(BaseModel):
    userMessage: str
    context: Dict[str, Any]


class UnderstandOutput(BaseModel):
    intent: str
    entities: Dict[str, Any]
    complexity: float
    ambiguity: float
    requiredActions: List[str]


class ReasonInput(BaseModel):
    understandOutput: UnderstandOutput
    context: Dict[str, Any]


class ReasonOutput(BaseModel):
    plan: Dict[str, Any]
    confidence: float


class ActInput(BaseModel):
    reasonOutput: ReasonOutput
    understandOutput: UnderstandOutput
    context: Dict[str, Any]


class ActOutput(BaseModel):
    response: str
    actions: List[Dict[str, Any]]
    version: str
    diagnostics: Dict[str, Any] | None = None


def complexity_score(input_payload: Dict[str, Any]) -> float:
    user_message = input_payload.get("userMessage", "")
    tokens = len(user_message.split())
    ambiguity = len([m for m in ["?", "or", "maybe", "perhaps", "might", "could", "possibly"] if m in user_message.lower()])
    required_actions = len([m for m in ["create", "schedule", "plan", "analyze", "reflect"] if m in user_message.lower()])
    return (tokens / 100) + (ambiguity * 0.5) + required_actions


async def understand(input_payload: Dict[str, Any]) -> Dict[str, Any]:
    complexity = complexity_score({"userMessage": input_payload.get("userMessage", "")})
    ambiguity = min(complexity * 0.2, 1)
    intent = "reflection"
    entities: Dict[str, Any] = {}
    required_actions = ["analyze", "respond"]
    output = UnderstandOutput(intent=intent, entities=entities, complexity=complexity, ambiguity=ambiguity, requiredActions=required_actions)
    return output.model_dump()


async def reason(input_payload: Dict[str, Any]) -> Dict[str, Any]:
    plan = {
        "steps": ["Analyze user intent", "Generate response", "Check coherence"],
        "reasoning": "Based on understanding phase, create a structured plan",
    }
    output = ReasonOutput(plan=plan, confidence=0.8)
    return output.model_dump()


async def act(input_payload: Dict[str, Any]) -> Dict[str, Any]:
    response = "This is a placeholder response from the cognition engine."
    actions: List[Dict[str, Any]] = []
    output = ActOutput(response=response, actions=actions, version="cognition.v1")
    return output.model_dump()


async def run_pipeline(input_payload: Dict[str, Any]) -> Dict[str, Any]:
    validated = UnderstandInput(**input_payload)
    understand_result = await understand(validated.model_dump())
    reason_input = ReasonInput(understandOutput=UnderstandOutput(**understand_result), context=validated.context)
    reason_result = await reason(reason_input.model_dump())
    act_input = ActInput(reasonOutput=ReasonOutput(**reason_result), understandOutput=UnderstandOutput(**understand_result), context=validated.context)
    act_result = await act(act_input.model_dump())
    return act_result

